src/sfizz/modulations/sources/Controller.cpp

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: BSD-2-Clause
+
+// This code is part of the sfizz library and is licensed under a BSD 2-clause
+// license. You should have receive a LICENSE.md file along with the code.
+// If not, contact the sfizz maintainers at https://github.com/sfztools/sfizz
+
+#include "Controller.h"
+#include "../ModKey.h"
+#include "../../Smoothers.h"
+#include "../../ModifierHelpers.h"
+#include "../../Resources.h"
+#include "../../Config.h"
+#include "../../Debug.h"
+#include <absl/container/flat_hash_map.h>
+
+namespace sfz {
+
+struct ControllerSource::Impl {
+    double sampleRate_ = config::defaultSampleRate;
+    Resources* res_ = nullptr;
+    absl::flat_hash_map<ModKey, Smoother> smoother_;
+};
+
+ControllerSource::ControllerSource(Resources& res)
+    : impl_(new Impl)
+{
+    impl_->res_ = &res;
```

Comment:
We could assert that the pointer is non-null, since this should be a reference but can't because of the oo structure.

Reply:
`res_` can be a reference if there exists the `Impl(Resources &)` ctor.
I don't see too much why this is a big deal though.

Reply:
I must have misplaced the comment I was thinking of something else and can't figure out what now. Nvm :)